### ISSUE TITLE
[multistage] Null Check Predicates Not Pushed Down

### ISSUE THREAD

**Issue by ankitsultana** [apache/pinot#10381]
Repro:

```
explain plan for select count(*) from userAttributes_OFFLINE WHERE (deviceOS is null)

Execution Plan
LogicalAggregate(group=[{}], EXPR$0=[$SUM0($0)])
  LogicalExchange(distribution=[hash])
    LogicalAggregate(group=[{}], EXPR$0=[COUNT()])
      LogicalValues(tuples=[[]])
```

```
explain plan for select count(*) from userAttributes_OFFLINE WHERE (deviceOS is null or deviceOS IN ('macos'))

Execution Plan
LogicalAggregate(group=[{}], EXPR$0=[$SUM0($0)])
  LogicalExchange(distribution=[hash])
    LogicalAggregate(group=[{}], EXPR$0=[COUNT()])
      LogicalProject(deviceOS=[$5])
        LogicalFilter(condition=[=($5, 'macos')])
          LogicalTableScan(table=[[userAttributes_OFFLINE]])
```

cc: @walterddr : Is this because we don't create types with nullability in `TypeFactory`?

related to: #10148

**Comment by walterddr** [comment]
correct. see https://github.com/apache/pinot/pull/10148/ which causes more problem than it solve

**Comment by ankitsultana** [comment]
What are the other problems though? You mean change in plan of some queries (like broadcast join being used for NOT IN)?

**Comment by walterddr** [comment]
here is a quick design doc: https://docs.google.com/document/d/16slt8TEfhJjDCjbE7K6za9hrB5WagiuKscFzod2Oib8/edit

TL;DR
similar to the POC in #10423 and following the comments from @ankitsultana and @Jackie-Jiang 
1. add `_isNullableField` field in `FieldSpec`, with a default value as false.
2. When resolving whether a column is nullable the migration transition plan is as follow
    - When column level nullability is not configured, default to false 
        - Matches current multi-stage engine behavior
        - For single-stage engine, behavior doesn’t change either by following query time option flag
    - When column level nullability is configured (explicitly to either true or false), only if the table level nullability is turned on the following applies
        - Multi-stage engine registers columns with explicitly true setting as nullable
        - Single-stage engine doesn’t change behavior

**Comment by walterddr** [comment]
CC @gortiz who is also interested in null handling work

**Comment by gortiz** [comment]
> add _isNullableField field in FieldSpec, with a default value as false.

That would imply that we can define nullability in 3 places:
1. For the whole column in tableConfig (with `tableIndexConfg.nullHandlingEnabled`)
2. For individual columns in tableConfig (with `fieldConfigList.indexes.null`).
3. For individual columns in schema (inside FieldSpec)

This is also related to issue https://github.com/apache/pinot/issues/10652 and PR https://github.com/apache/pinot/pull/10653, where given the problems we actually have with individual column nullability I proposed to remove option 2.

I think we need to discuss this in deep. Right now nullability is a table property. As far as I understand we all think it should be a column property (in the same table some columns may be nullable and some other columns may not). But I think we need to discuss:
1. How this affects V1 and V2
2. Where to define the nullability.

About 2. I also think we all are going to agree that nullability should be deterministically specified in a single place instead of having 2 or 3 options to do so.

**Comment by ankitsultana** [comment]
Thanks @walterddr for the doc.

From a user point of view, I feel we should keep columnLevelNullability as a new feature altogether.

So if column level nullability is disabled, things can stay as they are now.

If column level nullability is enabled, then we ignore table level nullability altogether, and both v1 and v2 engine should honor it in the same consistent way, which is:

1. During ingestion null vectors would be created based on column level nullability (this is independent of query engine anyways)
2. On query side both v1 and v2 engine would behave the same way when the null handling query option is enabled/disabled.
3. Null filters like IS NULL/IS NOT NULL should work consistently across the two engines.

**Comment by walterddr** [comment]
> Thanks @walterddr for the doc.
> 
> From a user point of view, I feel we should keep columnLevelNullability as a new feature altogether.

+1 on this point, but i would like to keep this issue solely discuss the impact of this new feature on query time.
> 
> So if column level nullability is disabled, things can stay as they are now.

"As they are now" is a bit loose, let me explain below:


> If column level nullability is enabled, then we ignore table level nullability altogether, and both v1 and v2 engine should honor it in the same consistent way, which is:

--> this is the part why i want this to stay as a query-time only flag
* regardless of column level nullability is enabled. we use table-nullability flag (which only used during ingestion time) for ingestion
* later we might use column level to control table level, but that's a separate issue all together

and query-time null can only be enabled if table has enabled null value ingestion flag.

> 
> 1. During ingestion null vectors would be created based on column level nullability (this is independent of query engine anyways)

--> see above, we do not do that for now, ingestion doesn't change for this new feature;

> 2. On query side both v1 and v2 engine would behave the same way when the null handling query option is enabled/disabled.

null handling is always enabled on v2 regardless, b/c there's no way to not handle null, say in a LEFT JOIN situation. so null is always enabled

> 3. Null filters like IS NULL/IS NOT NULL should work consistently across the two engines.

- currently IS_NULL and IS_NOT_NULL is not supported by v2, so the current behavior on v2 is as if the column is not nullable. which suites the default not-nullable behavior before adding this column level null flag --> if this is what we meant by "behavior doesn't change", then yes the behavior doesn't change (on v2)
- i have yet to check on v1 but as long as we dont use these settings in v1 we are fine. unless we want them to match

**Comment by ankitsultana** [comment]
Hey folks, I listed some scenarios in this doc: https://docs.google.com/document/d/12yCY1sfsu54fVS4Csmq3KrCp3yKv0qGOkluzlyIAHSs/edit

Goal is to see how results will look like for some example queries with a example column-level nullability config.

Please request comment access and I'll grant asap (don't have everyone's email).

**Comment by Jackie-Jiang** [comment]
We should discuss if v2 engine should rely on `enableNullHandling` query option. Currently all v2 operators always support processing `null`, and there are 2 options:
1. Leaf stage still rely on `enableNullHandling` query option from query
2. v2 engine always set `enableNullHandling` query option if it is not already set (so that we can still disable it in v1 if necessary)

**Comment by walterddr** [comment]
the discussion here is: the bottom line is this discussion should only be regarded to query time, no ingestion time behavior change is discussed here.

now with the query time:
config
===
- if QueryOption `enableNullhandling` flag is set, consider all columns to be nullable (exactly like current v1 behavior)
- if QueryOption is not set, use the schema level column `isNullableField` to consider whether it is nullable
    - this behavior should be honored by both v1 and v2  
    - for v2, enableNullHandling flag should not be used b/c it is not a runtime controllable feature 
        - tables are registered in catelog during server start time. At that time, column nullability is already loaded
        - we need a schema watcher to change table cache in case schema is changed but that's orthogonal to the discussion here

config interpretation (runtime behavior, only on v1)
===
if a column is configured nullable, we read the null vector -- if it doesn't exist, there's no null in this segment for this column
if a column is configured as non-nullable, we skip the null vector, event if it exist, and directly use the forward index value



The main discussion point here is whether we should introduce a "new query option" specifically for v2, say `enableNullHandlingIfColumnIsNullable` which
- if `enableNullHandling` is set (only on v1) keep current v1 behavior
- if `enableNullHandlingWhenColumnIsNullable` is set, honor schema column level nullability
- if nothing is set, all columns are non-nullable (which only affects runtime behavior, at v2 planning time column is still considered nullable, it will only affect some optimization rule, but final results will be the same

**Comment by gortiz** [comment]
I don't understand these properties. I would expect that in V2 there is no nullability query property. The engine should treat nullability as any other type property. The same we don't have properties to treat strings as binaries, we should not have properties to enable/disable nullability.

I think we have a clear way to define null column nullability: to set that in the schema. For compatibility reasons, we also have a way to define table level nullability in table config. We can treat the latter as the default column level nullability.

## enableNullhandling

Why do we need that? Ideally the table config nullability will tell us the default nullability. Do we expect users changing nullability at runtime? Why? That is not a normal flag in other databases.

Are you proposing this property in order to keep backward compatibility so if it is not enabled, V1 ignores column level nullability? My proposal is to do not have this. There is no backward compatibility issue given that column level nullability is a new feature.
1. If users do not define column level nullability in the schema, v1 will honor `tableConfig.getIndexingConfig().isNullHandlingEnabled `, which is backward compatible.
2. If users do define column level nullability, then they have changed the schema, so it is backward compatible to return different values.

## enableNullHandlingIfColumnIsNullable

In which situation we want this disabled? I mean, in which cases we expect users to define a column as nullable but want to skip that? It sounds super confusing to understand.

**Comment by walterddr** [comment]
> I don't understand these properties. I would expect that in V2 there is no nullability query property. The engine should treat nullability as any other type property. The same we don't have properties to treat strings as binaries, we should not have properties to enable/disable nullability.

Based on my proposal above (config) section, v2 doesn't need any nullability query property.

> I think we have a clear way to define null column nullability: to set that in the schema. For compatibility reasons, we also have a way to define table level nullability in table config. We can treat the latter as the default column level nullability.

Currently the "table-level nullability" is not "this table contains null", but an orthogonal property "for all the columns, generate null-vectors in the segment". this chart should show what I meant:


| Control | During Query | During Ingestion |
| -------- | ------- | ------- |
| Table-level Control | via `SET enableNullHandling=true;` | via `tableConfig.tableIndexConfig.nullHandlingEnabled` |
| Column-level Control | via schema `isNullableField` (discussed here) | via schema `isNullableField` (not discussed here) |

my argument: 
1. the `tableConfig.tableIndexConfig.nullHandlingEnabled` should not be used during query time (do not couple them)
2. `isNullableField` in schema can be coupled in the future but i dont want to couple them here and now due to backward compatibility discussions


> ## enableNullhandling
> Why do we need that? Ideally the table config nullability will tell us the default nullability. Do we expect users changing nullability at runtime? Why? That is not a normal flag in other databases.
> 
> Are you proposing this property in order to keep backward compatibility so if it is not enabled, V1 ignores column level nullability? My proposal is to do not have this. There is no backward compatibility issue given that column level nullability is a new feature.
> 
> 1. If users do not define column level nullability in the schema, v1 will honor `tableConfig.getIndexingConfig().isNullHandlingEnabled `, which is backward compatible.

IMO this is not a good idea, see my table above. we dont want to couple ingestion time config with query time config

> 2. If users do define column level nullability, then they have changed the schema, so it is backward compatible to return different values.
> 
> ## enableNullHandlingIfColumnIsNullable
> In which situation we want this disabled? I mean, in which cases we expect users to define a column as nullable but want to skip that? It sounds super confusing to understand.

this is what I meant in the config section above (sorry if it is not clear) - `enableNullHandlingIfColumnIsNullable` is a hidden query option auto-attached to the leaf stage of v2 only

**v1** behavior:
1. if `enableNullhandling` is set, current v1 behavior
2. if `enableNullhandling` is not set, current v1 behavior
3. if `enableNullHandlingIfColumnIsNullable` is set, utilize column-level nullability  (user wont see it when directly using v1)

**v2** behavior: 
1. always utilize column-level nullability, and auto attach `enableNullHandlingIfColumnIsNullable` option to leaf stage of v2.

i am not 100% sure `enableNullHandlingIfColumnIsNullable` is needed. if not (e.g. we can return different results from v1 when `enableNullhandling` is not set but column nullability is set), then the behavior is even simpler:

**v1** behavior:
1. if `enableNullhandling` is set, current v1 behavior
2. if `enableNullhandling` is not set, utilize column-level nullability  <-- different result

**v2** behavior: 
1. always utilize column-level nullability

**Comment by gortiz** [comment]
> IMO this is not a good idea, see my table above. we dont want to couple ingestion time config with query time config

I think that is where we see things differently. I think coupling these concepts simplify the problem a lot, making it easier to understand by devs but also Pinot users.

In my proposal, `column level nullability` is the source of truth. This is a concept independent of the config that we define as: `schema.col.isNullField == null || schema.col.isNullField == null && tableConfig.tableIndexConfig.nullHandlingEnabled == true`

I think we are saying more or less the same thing. The only differences are:
- In my proposal:
  - There is not `table level nullability`, there is a `default column nullability for this table`.
    - Therefore in my proposal `column level nullability` is always defined. Either explicitly by the schema or implicitly by the tableConfig.
  - V1 always needs to have `enableNullhandling`.
- In your proposal:
  - `column level nullability` is not defined if `schema.col.isNullField` is not explicitly added.
  - It is not defined what happens if `column level nullability` and `table level nullability` are both used in an incompatible way (ie a table has `nullHandlingEnabled` = true and column level `isNullField` is false).

**Comment by walterddr** [comment]
we can definitely discuss the ingestion level nullability however regardless of whether we use `column level nullability` as a source of truth for ingestion. the solution to the query side remains the same correct? 
e.g. 
- for query time, column level nullability is source of truth (in v2) regardless of explicit true/false or missing (default false) (it is always defined in the query sense)
- in v1 always needs to have enableNullhandling 

the main argument is that:
tableConfig should be ignored b/c it is an ingestion time config. it can be changed from one segment to another and only governs how the next segment is created, not the entire table, thus using it is problematic

**Comment by gortiz** [comment]
News here:
* https://github.com/apache/pinot/pull/11824 is being reviewed. This PR tries to bring column level nullability by changing the way null vector index is being read. If config indicates the index should not be read, null is returned and semantically V1 query engine behaves as there were no null (although the plan is not modified and therefore more expensive than when `SET enableNullHandling=false`.
* There is another PR (https://github.com/apache/pinot/pull/11935) that tries to implement the same feature but using another approach. Instead of changing whether the index is loaded or not, it adds a new query option that has to be enabled to use column level nullability. In case it is enabled, V1 will verify whether FieldSpec.isNullable() is false. In that case it won't use the null value vector. Again, plans are not changed. This solution is lighter in terms of semantic changes and the modification is simple (changing a boolean with an enum). But given the null handling code is very repetitive, it requires to apply a simple change in a lot of places, modifying more than 1k lines of code.
* I've modified both PRs to add a table showing the semantics. I think they make them easier to understand.
* @walterddr and I discussed about both PRs and considered that https://github.com/apache/pinot/pull/11935 is easier to understand, but we are open to other points of view
